src/storage/src/hummock/iterator/skip_watermark.rs

```diff
@@ -0,0 +1,400 @@
+                                            }
+                                        }
+                                        return false;
+                                    }
+                                }
+                            }
+                            WatermarkDirection::Descending => {
+                                return match inner_key.cmp(watermark.as_ref()) {
+                                    // Current key as not reached the watermark. Just return.
+                                    Ordering::Less | Ordering::Equal => false,
+                                    // Current key will be filtered by the watermark.
+                                    // Return true to further advance the key.
+                                    Ordering::Greater => true,
+                                };
+                            }
+                        }
+                    }
+                    Ordering::Greater => {
+                        return false;
+                    }
+                }
+            }
+        }
+        false
+    }
+
+    /// Advance the key until iterator invalid or the current key will not be filtered by the latest watermark.
+    /// Calling this method should ensure that the first remaining watermark has been advanced to the current key.
+    ///
+    /// Return a flag indicating whether should later advance the watermark.
+    async fn advance_key(&mut self) -> HummockResult<bool> {
+        if let Some((table_id, vnode, direction, watermark)) = self.remain_watermarks.front() {
+            while self.inner.is_valid() {
+                {
+                    let key = self.inner.key();
+                    let key_table_id = key.user_key.table_id;
+                    let (key_vnode, inner_key) = key.user_key.table_key.split_vnode();
+                    match (&key_table_id, &key_vnode).cmp(&(table_id, vnode)) {
+                        Ordering::Less => {
+                            return Ok(false);
+                        }
+                        Ordering::Equal => {
+                            if direction.filter_by_watermark(inner_key, watermark) {
+                                self.inner.next().await?;
+                            } else {
+                                return Ok(false);
+                            }
+                        }
+                        Ordering::Greater => {
+                            // The current key has advanced over the watermark.
+                            // We may return to advance the watermark before advancing the key.
+                            return Ok(true);
+                        }
+                    }
+                };
+            }
+        }
+        Ok(false)
+    }
+
+    async fn advance_key_and_watermark(&mut self) -> HummockResult<()> {
+        // advance key and watermark in an interleave manner until nothing
+        // changed after the method is called.
```

Comment:
should we need to monitor the `advance_key` during iteration?

Reply:
Since the advance key still needs to parse the key, I am considering whether we need to monitor the behavior of the iterator for debugging.